pkg/vsphere/toolbox/hgfs/server.go

```diff
@@ -228,6 +236,57 @@ func (s *Server) GetattrV2(p *Packet) (interface{}, error) {
 	return res, nil
 }
 
+// SetattrV2 handles OpSetattrV2 requests
+func (s *Server) SetattrV2(p *Packet) (interface{}, error) {
+	res := &ReplySetattrV2{}
+
+	req := new(RequestSetattrV2)
+	err := UnmarshalBinary(p.Payload, req)
+	if err != nil {
+		return nil, err
+	}
+
+	name := req.FileName.Path()
+
+	uid := -1
+	if req.Attr.Mask&AttrValidUserID == AttrValidUserID {
+		uid = int(req.Attr.UserID)
+	}
+
+	gid := -1
+	if req.Attr.Mask&AttrValidGroupID == AttrValidGroupID {
+		gid = int(req.Attr.GroupID)
+	}
+
+	err = s.chown(name, uid, gid)
+	if err != nil {
+		return nil, err
+	}
+
+	var perm os.FileMode
+
+	if req.Attr.Mask&AttrValidOwnerPerms == AttrValidOwnerPerms {
+		perm |= os.FileMode(req.Attr.OwnerPerms) << 6
+	}
+
+	if req.Attr.Mask&AttrValidGroupPerms == AttrValidGroupPerms {
+		perm |= os.FileMode(req.Attr.GroupPerms) << 3
+	}
+
+	if req.Attr.Mask&AttrValidOtherPerms == AttrValidOtherPerms {
+		perm |= os.FileMode(req.Attr.OtherPerms)
+	}
+
+	if perm != 0 {
+		err = s.chmod(name, perm)
```

Comment:
Is it possible to extract this information before we start writing the file? Otherwise there's a period of time where the permissions and uid/gid do not align with the attrs.
If not we should likely make sure there's an exclusive lock until we have applied said attrs.